// @nguniversal/socket-engine/esm2015/src/main.js
/**
 * @fileoverview added by tsickle
 * Generated from: modules/socket-engine/src/main.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { __awaiter } from "tslib";
import { ɵCommonEngine as CommonEngine } from '@nguniversal/common/engine';
import * as net from 'net';
/**
 * @record
 */
export function SocketEngineServer() { }
if (false) {
    /** @type {?} */
    SocketEngineServer.prototype.close;
}
/**
 * @record
 */
export function SocketEngineRenderOptions() { }
if (false) {
    /** @type {?} */
    SocketEngineRenderOptions.prototype.id;
}
/**
 * @record
 */
export function SocketEngineResponse() { }
if (false) {
    /** @type {?} */
    SocketEngineResponse.prototype.id;
    /** @type {?} */
    SocketEngineResponse.prototype.html;
    /** @type {?|undefined} */
    SocketEngineResponse.prototype.error;
}
/**
 * @param {?} moduleOrFactory
 * @param {?=} providers
 * @param {?=} host
 * @param {?=} port
 * @return {?}
 * @this {*}
 */
export function startSocketEngine(moduleOrFactory, providers = [], host = 'localhost', port = 9090) {
    return new Promise((/**
     * @param {?} resolve
     * @param {?} _reject
     * @return {?}
     */
    (resolve, _reject) => {
        /** @type {?} */
        const engine = new CommonEngine(moduleOrFactory, providers);
        /** @type {?} */
        const server = net.createServer((/**
         * @param {?} socket
         * @return {?}
         */
        socket => {
            socket.on('data', (/**
             * @param {?} buff
             * @return {?}
             */
            (buff) => __awaiter(this, void 0, void 0, function* () {
                /** @type {?} */
                const message = buff.toString();
                /** @type {?} */
                const renderOptions = (/** @type {?} */ (JSON.parse(message)));
                try {
                    /** @type {?} */
                    const html = yield engine.render(renderOptions);
                    socket.write(JSON.stringify((/** @type {?} */ ({ html, id: renderOptions.id }))));
                }
                catch (error) {
                    // send the error down to the client then rethrow it
                    socket.write(JSON.stringify((/** @type {?} */ ({
                        html: null,
                        id: renderOptions.id,
                        error: error.toString(),
                    }))));
                }
            })));
        }));
        server.listen(port, host);
        resolve({ close: (/**
             * @return {?}
             */
            () => server.close()) });
    }));
}
