// @locl/cli/src/cmds/extract/extractor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@babel/core");
const es2015_extract_plugin_1 = require("./source_files/es2015_extract_plugin");
const es5_extract_plugin_1 = require("./source_files/es5_extract_plugin");
/**
 * A class that extracts parsed messages from file contents, by parsing the contents as JavaScript
 * and looking for occurrences of `$localize` in the source code.
 */
class Extractor {
    constructor(diagnostics) {
        this.diagnostics = diagnostics;
        this.messages = [];
    }
    extractMessages(sourceCode) {
        if (sourceCode.includes('$localize')) {
            // Only bother to parse the file if it contains a reference to `$localize`.
            core_1.transformSync(sourceCode, {
                plugins: [
                    es2015_extract_plugin_1.makeEs2015ExtractPlugin(this.messages, this.diagnostics),
                    es5_extract_plugin_1.makeEs5ExtractPlugin(this.messages, this.diagnostics)
                ],
                code: false,
                ast: false
            });
        }
    }
}
exports.Extractor = Extractor;
