// @danielmoncada/angular-datetime-picker/esm2015/lib/date-time/date-time-picker-trigger.directive.js
/**
 * date-time-picker-trigger.directive
 */
import { __decorate, __metadata } from "tslib";
import { AfterContentInit, ChangeDetectorRef, Directive, Input, OnChanges, OnDestroy, OnInit, SimpleChanges } from '@angular/core';
import { OwlDateTimeComponent } from './date-time-picker.component';
import { merge, of as observableOf, Subscription } from 'rxjs';
let OwlDateTimeTriggerDirective = class OwlDateTimeTriggerDirective {
    constructor(changeDetector) {
        this.changeDetector = changeDetector;
        this.stateChanges = Subscription.EMPTY;
    }
    get disabled() {
        return this._disabled === undefined ? this.dtPicker.disabled : !!this._disabled;
    }
    set disabled(value) {
        this._disabled = value;
    }
    get owlDTTriggerDisabledClass() {
        return this.disabled;
    }
    ngOnInit() {
    }
    ngOnChanges(changes) {
        if (changes.datepicker) {
            this.watchStateChanges();
        }
    }
    ngAfterContentInit() {
        this.watchStateChanges();
    }
    ngOnDestroy() {
        this.stateChanges.unsubscribe();
    }
    handleClickOnHost(event) {
        if (this.dtPicker) {
            this.dtPicker.open();
            event.stopPropagation();
        }
    }
    watchStateChanges() {
        this.stateChanges.unsubscribe();
        const inputDisabled = this.dtPicker && this.dtPicker.dtInput ?
            this.dtPicker.dtInput.disabledChange : observableOf();
        const pickerDisabled = this.dtPicker ?
            this.dtPicker.disabledChange : observableOf();
        this.stateChanges = merge(pickerDisabled, inputDisabled)
            .subscribe(() => {
            this.changeDetector.markForCheck();
        });
    }
};
OwlDateTimeTriggerDirective.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
__decorate([
    Input('owlDateTimeTrigger'),
    __metadata("design:type", OwlDateTimeComponent)
], OwlDateTimeTriggerDirective.prototype, "dtPicker", void 0);
__decorate([
    Input(),
    __metadata("design:type", Boolean),
    __metadata("design:paramtypes", [Boolean])
], OwlDateTimeTriggerDirective.prototype, "disabled", null);
OwlDateTimeTriggerDirective = __decorate([
    Directive({
        selector: '[owlDateTimeTrigger]',
        host: {
            '(click)': 'handleClickOnHost($event)',
            '[class.owl-dt-trigger-disabled]': 'owlDTTriggerDisabledClass'
        }
    }),
    __metadata("design:paramtypes", [ChangeDetectorRef])
], OwlDateTimeTriggerDirective);
export { OwlDateTimeTriggerDirective };
