// @locl/cli/src/cmds/convert/translation_parsers/simple_json_translation_parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
const localize_1 = require("@angular/localize");
const path_1 = require("path");
/**
 * A translation parser that can parse JSON that has the form:
 *
 * ```
 * {
 *   "locale": "...",
 *   "translations": {
 *     "message-id": "Target message string",
 *     ...
 *   }
 * }
 * ```
 */
class SimpleJsonTranslationParser {
    constructor(diagnostics) {
        this.diagnostics = diagnostics;
    }
    canParse(filePath, _contents) {
        return path_1.extname(filePath) === '.json';
    }
    parse(_filePath, contents) {
        const { locale: parsedLocale, translations } = JSON.parse(contents);
        const parsedTranslations = {};
        for (const messageId in translations) {
            const targetMessage = translations[messageId];
            parsedTranslations[messageId] = localize_1.ɵparseTranslation(targetMessage);
        }
        return {
            locale: parsedLocale,
            translations: parsedTranslations,
            diagnostics: this.diagnostics
        };
    }
}
exports.SimpleJsonTranslationParser = SimpleJsonTranslationParser;
