// @locl/cli/src/cmds/convert/translation_parsers/translation_parse_error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
const compiler_1 = require("@angular/compiler");
/**
 * This error is thrown when there is a problem parsing a translation file.
 */
class TranslationParseError extends Error {
    constructor(span, msg, level = compiler_1.ParseErrorLevel.ERROR) {
        super(msg);
        this.span = span;
        this.msg = msg;
        this.level = level;
    }
    contextualMessage() {
        const ctx = this.span.start.getContext(100, 3);
        return ctx
            ? `${this.msg} ("${ctx.before}[${compiler_1.ParseErrorLevel[this.level]} ->]${ctx.after}")`
            : this.msg;
    }
    toString() {
        const details = this.span.details ? `, ${this.span.details}` : '';
        return `${this.contextualMessage()}: ${this.span.start}${details}`;
    }
}
exports.TranslationParseError = TranslationParseError;
