// @locl/cli/esm5/cmds/extract/translation_files/xliff1_translation_serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: cmds/extract/translation_files/xliff1_translation_serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { XmlFile } from './xml_file';
var Xliff1TranslationSerializer = /** @class */ (function () {
    function Xliff1TranslationSerializer() {
    }
    /**
     * @param {?} messages
     * @param {?} locale
     * @param {?=} isTarget
     * @return {?}
     */
    Xliff1TranslationSerializer.prototype.renderFile = /**
     * @param {?} messages
     * @param {?} locale
     * @param {?=} isTarget
     * @return {?}
     */
    function (messages, locale, isTarget) {
        var _a;
        var _this = this;
        if (isTarget === void 0) { isTarget = false; }
        /** @type {?} */
        var xml = new XmlFile();
        xml.startTag('xliff', {
            version: '1.2',
            xmlns: 'urn:oasis:names:tc:xliff:document:1.2'
        });
        /** @type {?} */
        var fileAttrs = (_a = {},
            _a['source-language'] = isTarget ? undefined : locale,
            _a['target-language'] = locale,
            _a.datatype = 'plaintext',
            _a);
        xml.startTag('file', fileAttrs);
        xml.startTag('body');
        messages.forEach((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            xml.startTag('trans-unit', {
                id: ((/** @type {?} */ (message))).id ||
                    ((/** @type {?} */ (message))).messageId,
                datatype: 'html'
            });
            if (!isTarget) {
                _this.generateMessageTag(xml, 'source', message);
            }
            _this.generateMessageTag(xml, 'target', message);
            if (message.description) {
                _this.renderNote(xml, 'description', message.description);
            }
            if (message.meaning) {
                _this.renderNote(xml, 'meaning', message.meaning);
            }
            xml.endTag('trans-unit');
        }));
        xml.endTag('body');
        xml.endTag('file');
        xml.endTag('xliff');
        return xml.toString();
    };
    /**
     * @private
     * @param {?} xml
     * @param {?} tagName
     * @param {?} message
     * @return {?}
     */
    Xliff1TranslationSerializer.prototype.generateMessageTag = /**
     * @private
     * @param {?} xml
     * @param {?} tagName
     * @param {?} message
     * @return {?}
     */
    function (xml, tagName, message) {
        xml.startTag(tagName, {}, { preserveWhitespace: true });
        this.renderMessage(xml, message);
        xml.endTag(tagName, { preserveWhitespace: false });
    };
    /**
     * @private
     * @param {?} xml
     * @param {?} message
     * @return {?}
     */
    Xliff1TranslationSerializer.prototype.renderMessage = /**
     * @private
     * @param {?} xml
     * @param {?} message
     * @return {?}
     */
    function (xml, message) {
        xml.text(message.messageParts[0]);
        for (var i = 1; i < message.messageParts.length; i++) {
            xml.startTag('x', { id: message.placeholderNames[i - 1] }, { selfClosing: true });
            xml.text(message.messageParts[i]);
        }
    };
    /**
     * @private
     * @param {?} xml
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    Xliff1TranslationSerializer.prototype.renderNote = /**
     * @private
     * @param {?} xml
     * @param {?} name
     * @param {?} value
     * @return {?}
     */
    function (xml, name, value) {
        xml.startTag('note', { priority: '1', from: name }, { preserveWhitespace: true });
        xml.text(value);
        xml.endTag('note', { preserveWhitespace: false });
    };
    return Xliff1TranslationSerializer;
}());
export { Xliff1TranslationSerializer };
