// @locl/cli/esm5/cmds/extract/translation_files/xmb_translation_serializer.js
/**
 * @fileoverview added by tsickle
 * Generated from: cmds/extract/translation_files/xmb_translation_serializer.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { XmlFile } from './xml_file';
/** @type {?} */
var DOCTYPE = "<!ELEMENT messagebundle (msg)*>\n<!ATTLIST messagebundle class CDATA #IMPLIED>\n\n<!ELEMENT msg (#PCDATA|ph|source)*>\n<!ATTLIST msg id CDATA #IMPLIED>\n<!ATTLIST msg seq CDATA #IMPLIED>\n<!ATTLIST msg name CDATA #IMPLIED>\n<!ATTLIST msg desc CDATA #IMPLIED>\n<!ATTLIST msg meaning CDATA #IMPLIED>\n<!ATTLIST msg obsolete (obsolete) #IMPLIED>\n<!ATTLIST msg xml:space (default|preserve) \"default\">\n<!ATTLIST msg is_hidden CDATA #IMPLIED>\n\n<!ELEMENT source (#PCDATA)>\n\n<!ELEMENT ph (#PCDATA|ex)*>\n<!ATTLIST ph name CDATA #REQUIRED>\n\n<!ELEMENT ex (#PCDATA)>";
var XmbTranslationSerializer = /** @class */ (function () {
    function XmbTranslationSerializer() {
    }
    /**
     * @param {?} messages
     * @return {?}
     */
    XmbTranslationSerializer.prototype.renderFile = /**
     * @param {?} messages
     * @return {?}
     */
    function (messages) {
        var _this = this;
        /** @type {?} */
        var xml = new XmlFile();
        xml.startTag('messagebundle');
        messages.forEach((/**
         * @param {?} message
         * @return {?}
         */
        function (message) {
            xml.startTag('msg', {
                id: ((/** @type {?} */ (message))).id ||
                    ((/** @type {?} */ (message))).messageId,
                desc: message.description,
                meaning: message.meaning
            }, { preserveWhitespace: true });
            _this.renderMessage(xml, message);
            xml.endTag('msg', { preserveWhitespace: false });
        }));
        xml.endTag('messagebundle');
        return xml.toString();
    };
    /**
     * @private
     * @param {?} xml
     * @param {?} message
     * @return {?}
     */
    XmbTranslationSerializer.prototype.renderMessage = /**
     * @private
     * @param {?} xml
     * @param {?} message
     * @return {?}
     */
    function (xml, message) {
        xml.text(message.messageParts[0]);
        for (var i = 1; i < message.messageParts.length; i++) {
            xml.startTag('ph', { name: message.placeholderNames[i - 1] }, { selfClosing: true });
            xml.text(message.messageParts[i]);
        }
    };
    return XmbTranslationSerializer;
}());
export { XmbTranslationSerializer };
