// @locl/cli/esm5/cmds/extract/translation_files/xml_file.js
/**
 * @fileoverview added by tsickle
 * Generated from: cmds/extract/translation_files/xml_file.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * @record
 */
function Options() { }
if (false) {
    /** @type {?|undefined} */
    Options.prototype.selfClosing;
    /** @type {?|undefined} */
    Options.prototype.preserveWhitespace;
}
var XmlFile = /** @class */ (function () {
    function XmlFile() {
        this.output = '';
        this.indent = '';
        this.elements = [];
        this.preservingWhitespace = false;
    }
    /**
     * @return {?}
     */
    XmlFile.prototype.toString = /**
     * @return {?}
     */
    function () {
        return this.output;
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} name
     * @param {?=} attributes
     * @param {?=} __2
     * @return {THIS}
     */
    XmlFile.prototype.startTag = /**
     * @template THIS
     * @this {THIS}
     * @param {?} name
     * @param {?=} attributes
     * @param {?=} __2
     * @return {THIS}
     */
    function (name, attributes, _a) {
        var _this = this;
        if (attributes === void 0) { attributes = {}; }
        var _b = _a === void 0 ? {} : _a, _c = _b.selfClosing, selfClosing = _c === void 0 ? false : _c, preserveWhitespace = _b.preserveWhitespace;
        if (!(/** @type {?} */ (this)).preservingWhitespace) {
            (/** @type {?} */ (this)).output += (/** @type {?} */ (this)).indent;
        }
        (/** @type {?} */ (this)).output += "<" + name;
        Object.keys(attributes).forEach((/**
         * @param {?} attrName
         * @return {?}
         */
        function (attrName) {
            if (attributes[attrName]) {
                (/** @type {?} */ (_this)).output += " " + attrName + "=\"" + escapeXml(attributes[attrName]) + "\"";
            }
        }));
        if (selfClosing) {
            (/** @type {?} */ (this)).output += '/>';
        }
        else {
            (/** @type {?} */ (this)).output += '>';
            (/** @type {?} */ (this)).elements.push(name);
            (/** @type {?} */ (this)).incIndent();
        }
        if (preserveWhitespace !== undefined) {
            (/** @type {?} */ (this)).preservingWhitespace = preserveWhitespace;
        }
        if (!(/** @type {?} */ (this)).preservingWhitespace) {
            (/** @type {?} */ (this)).output += "\n";
        }
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} name
     * @param {?=} __1
     * @return {THIS}
     */
    XmlFile.prototype.endTag = /**
     * @template THIS
     * @this {THIS}
     * @param {?} name
     * @param {?=} __1
     * @return {THIS}
     */
    function (name, _a) {
        var preserveWhitespace = (_a === void 0 ? {} : _a).preserveWhitespace;
        /** @type {?} */
        var expectedTag = (/** @type {?} */ (this)).elements.pop();
        if (expectedTag !== name) {
            throw new Error("Unexpected closing tag: \"" + name + "\", expected: \"" + expectedTag + "\"");
        }
        (/** @type {?} */ (this)).decIndent();
        if (!(/** @type {?} */ (this)).preservingWhitespace) {
            (/** @type {?} */ (this)).output += (/** @type {?} */ (this)).indent;
        }
        (/** @type {?} */ (this)).output += "</" + name + ">";
        if (preserveWhitespace !== undefined) {
            (/** @type {?} */ (this)).preservingWhitespace = preserveWhitespace;
        }
        if (!(/** @type {?} */ (this)).preservingWhitespace) {
            (/** @type {?} */ (this)).output += "\n";
        }
        return (/** @type {?} */ (this));
    };
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} str
     * @return {THIS}
     */
    XmlFile.prototype.text = /**
     * @template THIS
     * @this {THIS}
     * @param {?} str
     * @return {THIS}
     */
    function (str) {
        (/** @type {?} */ (this)).output += escapeXml(str);
        return (/** @type {?} */ (this));
    };
    /**
     * @return {?}
     */
    XmlFile.prototype.startPreserveWhitespace = /**
     * @return {?}
     */
    function () {
        this.preservingWhitespace = true;
    };
    /**
     * @return {?}
     */
    XmlFile.prototype.stopPreserveWhitespace = /**
     * @return {?}
     */
    function () {
        this.preservingWhitespace = false;
    };
    /**
     * @private
     * @return {?}
     */
    XmlFile.prototype.incIndent = /**
     * @private
     * @return {?}
     */
    function () {
        this.indent = this.indent + '  ';
    };
    /**
     * @private
     * @return {?}
     */
    XmlFile.prototype.decIndent = /**
     * @private
     * @return {?}
     */
    function () {
        this.indent = this.indent.slice(0, -2);
    };
    return XmlFile;
}());
export { XmlFile };
if (false) {
    /**
     * @type {?}
     * @private
     */
    XmlFile.prototype.output;
    /**
     * @type {?}
     * @private
     */
    XmlFile.prototype.indent;
    /**
     * @type {?}
     * @private
     */
    XmlFile.prototype.elements;
    /**
     * @type {?}
     * @private
     */
    XmlFile.prototype.preservingWhitespace;
}
/** @type {?} */
var _ESCAPED_CHARS = [
    [/&/g, '&amp;'],
    [/"/g, '&quot;'],
    [/'/g, '&apos;'],
    [/</g, '&lt;'],
    [/>/g, '&gt;']
];
/**
 * @param {?} text
 * @return {?}
 */
function escapeXml(text) {
    return _ESCAPED_CHARS.reduce((/**
     * @param {?} t
     * @param {?} entry
     * @return {?}
     */
    function (t, entry) { return t.replace(entry[0], entry[1]); }), text);
}
