// @locl/cli/esm2015/cmds/extract/translation_files/xml_file.js
/**
 * @fileoverview added by tsickle
 * Generated from: cmds/extract/translation_files/xml_file.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
/**
 * @record
 */
function Options() { }
if (false) {
    /** @type {?|undefined} */
    Options.prototype.selfClosing;
    /** @type {?|undefined} */
    Options.prototype.preserveWhitespace;
}
export class XmlFile {
    constructor() {
        this.output = '';
        this.indent = '';
        this.elements = [];
        this.preservingWhitespace = false;
    }
    /**
     * @return {?}
     */
    toString() {
        return this.output;
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} name
     * @param {?=} attributes
     * @param {?=} __2
     * @return {THIS}
     */
    startTag(name, attributes = {}, { selfClosing = false, preserveWhitespace } = {}) {
        if (!(/** @type {?} */ (this)).preservingWhitespace) {
            (/** @type {?} */ (this)).output += (/** @type {?} */ (this)).indent;
        }
        (/** @type {?} */ (this)).output += `<${name}`;
        Object.keys(attributes).forEach((/**
         * @param {?} attrName
         * @return {?}
         */
        attrName => {
            if (attributes[attrName]) {
                (/** @type {?} */ (this)).output += ` ${attrName}="${escapeXml(attributes[attrName])}"`;
            }
        }));
        if (selfClosing) {
            (/** @type {?} */ (this)).output += '/>';
        }
        else {
            (/** @type {?} */ (this)).output += '>';
            (/** @type {?} */ (this)).elements.push(name);
            (/** @type {?} */ (this)).incIndent();
        }
        if (preserveWhitespace !== undefined) {
            (/** @type {?} */ (this)).preservingWhitespace = preserveWhitespace;
        }
        if (!(/** @type {?} */ (this)).preservingWhitespace) {
            (/** @type {?} */ (this)).output += `\n`;
        }
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} name
     * @param {?=} __1
     * @return {THIS}
     */
    endTag(name, { preserveWhitespace } = {}) {
        /** @type {?} */
        const expectedTag = (/** @type {?} */ (this)).elements.pop();
        if (expectedTag !== name) {
            throw new Error(`Unexpected closing tag: "${name}", expected: "${expectedTag}"`);
        }
        (/** @type {?} */ (this)).decIndent();
        if (!(/** @type {?} */ (this)).preservingWhitespace) {
            (/** @type {?} */ (this)).output += (/** @type {?} */ (this)).indent;
        }
        (/** @type {?} */ (this)).output += `</${name}>`;
        if (preserveWhitespace !== undefined) {
            (/** @type {?} */ (this)).preservingWhitespace = preserveWhitespace;
        }
        if (!(/** @type {?} */ (this)).preservingWhitespace) {
            (/** @type {?} */ (this)).output += `\n`;
        }
        return (/** @type {?} */ (this));
    }
    /**
     * @template THIS
     * @this {THIS}
     * @param {?} str
     * @return {THIS}
     */
    text(str) {
        (/** @type {?} */ (this)).output += escapeXml(str);
        return (/** @type {?} */ (this));
    }
    /**
     * @return {?}
     */
    startPreserveWhitespace() {
        this.preservingWhitespace = true;
    }
    /**
     * @return {?}
     */
    stopPreserveWhitespace() {
        this.preservingWhitespace = false;
    }
    /**
     * @private
     * @return {?}
     */
    incIndent() {
        this.indent = this.indent + '  ';
    }
    /**
     * @private
     * @return {?}
     */
    decIndent() {
        this.indent = this.indent.slice(0, -2);
    }
}
if (false) {
    /**
     * @type {?}
     * @private
     */
    XmlFile.prototype.output;
    /**
     * @type {?}
     * @private
     */
    XmlFile.prototype.indent;
    /**
     * @type {?}
     * @private
     */
    XmlFile.prototype.elements;
    /**
     * @type {?}
     * @private
     */
    XmlFile.prototype.preservingWhitespace;
}
/** @type {?} */
const _ESCAPED_CHARS = [
    [/&/g, '&amp;'],
    [/"/g, '&quot;'],
    [/'/g, '&apos;'],
    [/</g, '&lt;'],
    [/>/g, '&gt;']
];
/**
 * @param {?} text
 * @return {?}
 */
function escapeXml(text) {
    return _ESCAPED_CHARS.reduce((/**
     * @param {?} t
     * @param {?} entry
     * @return {?}
     */
    (t, entry) => t.replace(entry[0], entry[1])), text);
}
