// @locl/cli/esm5/cmds/common/diagnostics.js
/**
 * @fileoverview added by tsickle
 * Generated from: cmds/common/diagnostics.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as chalk from 'chalk';
/**
 * This class is used to collect and then report warnings and errors that occur during the execution
 * of the tools.
 */
var /**
 * This class is used to collect and then report warnings and errors that occur during the execution
 * of the tools.
 */
Diagnostics = /** @class */ (function () {
    function Diagnostics() {
        this.messages = [];
    }
    Object.defineProperty(Diagnostics.prototype, "hasErrors", {
        get: /**
         * @return {?}
         */
        function () {
            return this.messages.some((/**
             * @param {?} m
             * @return {?}
             */
            function (m) { return m.type === 'error'; }));
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @param {?} message
     * @return {?}
     */
    Diagnostics.prototype.log = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this.messages.push({ type: '', message: message });
    };
    /**
     * @param {?} message
     * @return {?}
     */
    Diagnostics.prototype.warn = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this.messages.push({ type: 'warning', message: message });
    };
    /**
     * @param {?} message
     * @return {?}
     */
    Diagnostics.prototype.error = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        this.messages.push({ type: 'error', message: message });
    };
    /**
     * @param {?} message
     * @return {?}
     */
    Diagnostics.prototype.formatDiagnostics = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        /** @type {?} */
        var errors = (/** @type {?} */ (this.messages)).filter((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.type === 'error'; })).map((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return ' - ' + d.message; }));
        /** @type {?} */
        var warnings = (/** @type {?} */ (this.messages)).filter((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return d.type === 'warning'; })).map((/**
         * @param {?} d
         * @return {?}
         */
        function (d) { return ' - ' + d.message; }));
        if (errors.length) {
            message += '\nERRORS:\n' + errors.join('\n');
        }
        if (warnings.length) {
            message += '\nWARNINGS:\n' + warnings.join('\n');
        }
        return message;
    };
    /**
     * @return {?}
     */
    Diagnostics.prototype.logMessages = /**
     * @return {?}
     */
    function () {
        while (this.messages.length) {
            /** @type {?} */
            var m = this.messages.shift();
            switch (m.type) {
                case 'warning':
                    console.warn(chalk.yellow("Warning: " + m.message));
                    break;
                case 'error':
                    console.error(chalk.red("Error: " + m.message));
                    break;
                default:
                    console.log(chalk.blue("" + m.message));
            }
        }
    };
    return Diagnostics;
}());
/**
 * This class is used to collect and then report warnings and errors that occur during the execution
 * of the tools.
 */
export { Diagnostics };
if (false) {
    /** @type {?} */
    Diagnostics.prototype.messages;
}
