// @locl/cli/esm2015/cmds/common/diagnostics.js
/**
 * @fileoverview added by tsickle
 * Generated from: cmds/common/diagnostics.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as chalk from 'chalk';
/**
 * This class is used to collect and then report warnings and errors that occur during the execution
 * of the tools.
 */
export class Diagnostics {
    constructor() {
        this.messages = [];
    }
    /**
     * @return {?}
     */
    get hasErrors() {
        return this.messages.some((/**
         * @param {?} m
         * @return {?}
         */
        m => m.type === 'error'));
    }
    /**
     * @param {?} message
     * @return {?}
     */
    log(message) {
        this.messages.push({ type: '', message });
    }
    /**
     * @param {?} message
     * @return {?}
     */
    warn(message) {
        this.messages.push({ type: 'warning', message });
    }
    /**
     * @param {?} message
     * @return {?}
     */
    error(message) {
        this.messages.push({ type: 'error', message });
    }
    /**
     * @param {?} message
     * @return {?}
     */
    formatDiagnostics(message) {
        /** @type {?} */
        const errors = (/** @type {?} */ (this.messages)).filter((/**
         * @param {?} d
         * @return {?}
         */
        d => d.type === 'error')).map((/**
         * @param {?} d
         * @return {?}
         */
        d => ' - ' + d.message));
        /** @type {?} */
        const warnings = (/** @type {?} */ (this.messages)).filter((/**
         * @param {?} d
         * @return {?}
         */
        d => d.type === 'warning')).map((/**
         * @param {?} d
         * @return {?}
         */
        d => ' - ' + d.message));
        if (errors.length) {
            message += '\nERRORS:\n' + errors.join('\n');
        }
        if (warnings.length) {
            message += '\nWARNINGS:\n' + warnings.join('\n');
        }
        return message;
    }
    /**
     * @return {?}
     */
    logMessages() {
        while (this.messages.length) {
            /** @type {?} */
            const m = this.messages.shift();
            switch (m.type) {
                case 'warning':
                    console.warn(chalk.yellow(`Warning: ${m.message}`));
                    break;
                case 'error':
                    console.error(chalk.red(`Error: ${m.message}`));
                    break;
                default:
                    console.log(chalk.blue(`${m.message}`));
            }
        }
    }
}
if (false) {
    /** @type {?} */
    Diagnostics.prototype.messages;
}
