// @locl/cli/esm2015/cmds/common/file_utils.js
/**
 * @fileoverview added by tsickle
 * Generated from: cmds/common/file_utils.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google Inc. All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import * as fs from 'fs';
import * as path from 'path';
export class FileUtils {
    /**
     * @param {?} absolutePath
     * @return {?}
     */
    static readFile(absolutePath) {
        return fs.readFileSync(absolutePath, 'utf8');
    }
    /**
     * @param {?} absolutePath
     * @return {?}
     */
    static readFileBuffer(absolutePath) {
        return fs.readFileSync(absolutePath);
    }
    /**
     * @param {?} absolutePath
     * @param {?} contents
     * @return {?}
     */
    static writeFile(absolutePath, contents) {
        FileUtils.ensureDir(path.dirname(absolutePath));
        fs.writeFileSync(absolutePath, contents);
    }
    /**
     * @param {?} absolutePath
     * @return {?}
     */
    static ensureDir(absolutePath) {
        /** @type {?} */
        const parents = [];
        while (!FileUtils.isRoot(absolutePath) && !fs.existsSync(absolutePath)) {
            parents.push(absolutePath);
            absolutePath = path.dirname(absolutePath);
        }
        while (parents.length) {
            fs.mkdirSync((/** @type {?} */ (parents.pop())));
        }
    }
    /**
     * @param {?} p
     * @return {?}
     */
    static remove(p) {
        /** @type {?} */
        const stat = fs.statSync(p);
        if (stat.isFile()) {
            fs.unlinkSync(p);
        }
        else if (stat.isDirectory()) {
            fs.readdirSync(p).forEach((/**
             * @param {?} child
             * @return {?}
             */
            child => {
                /** @type {?} */
                const absChild = path.resolve(p, child);
                FileUtils.remove(absChild);
            }));
            fs.rmdirSync(p);
        }
    }
    /**
     * @param {?} absolutePath
     * @return {?}
     */
    static isRoot(absolutePath) {
        return path.dirname(absolutePath) === absolutePath;
    }
    /**
     * @param {?} files
     * @param {?} pattern
     * @param {?=} replaceValue
     * @return {?}
     */
    static dedup(files, pattern, replaceValue = '') {
        /** @type {?} */
        const filesSet = files.map((/**
         * @param {?} file
         * @return {?}
         */
        file => file.replace(pattern, replaceValue)));
        /** @type {?} */
        const dedup = [];
        /** @type {?} */
        const indexes = [];
        filesSet.forEach((/**
         * @param {?} file
         * @param {?} index
         * @return {?}
         */
        (file, index) => {
            if (dedup.indexOf(file) === -1) {
                dedup.push(file);
                indexes.push(index);
            }
        }));
        return indexes.map((/**
         * @param {?} index
         * @return {?}
         */
        index => files[index]));
    }
}
