// @locl/cli/esm5/cmds/extract/extractor.js
/**
 * @fileoverview added by tsickle
 * Generated from: cmds/extract/extractor.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { transformSync } from '@babel/core';
import { makeEs2015ExtractPlugin } from './source_files/es2015_extract_plugin';
import { makeEs5ExtractPlugin } from './source_files/es5_extract_plugin';
/**
 * A class that extracts parsed messages from file contents, by parsing the contents as JavaScript
 * and looking for occurrences of `$localize` in the source code.
 */
var /**
 * A class that extracts parsed messages from file contents, by parsing the contents as JavaScript
 * and looking for occurrences of `$localize` in the source code.
 */
Extractor = /** @class */ (function () {
    function Extractor(diagnostics) {
        this.diagnostics = diagnostics;
        this.messages = [];
    }
    /**
     * @param {?} sourceCode
     * @return {?}
     */
    Extractor.prototype.extractMessages = /**
     * @param {?} sourceCode
     * @return {?}
     */
    function (sourceCode) {
        if (sourceCode.includes('$localize')) {
            // Only bother to parse the file if it contains a reference to `$localize`.
            transformSync(sourceCode, {
                plugins: [
                    makeEs2015ExtractPlugin(this.messages, this.diagnostics),
                    makeEs5ExtractPlugin(this.messages, this.diagnostics)
                ],
                code: false,
                ast: false
            });
        }
    };
    return Extractor;
}());
/**
 * A class that extracts parsed messages from file contents, by parsing the contents as JavaScript
 * and looking for occurrences of `$localize` in the source code.
 */
export { Extractor };
if (false) {
    /** @type {?} */
    Extractor.prototype.messages;
    /**
     * @type {?}
     * @private
     */
    Extractor.prototype.diagnostics;
}
